# Grouping for Aggregation, Filtration and Transformation

In [3]:
import pandas as pd
import numpy as np

## Defining an aggregation

In [2]:
flights = pd.read_csv('data/flights.csv')
flights.head()

MONTH  DAY  WEEKDAY AIRLINE ORG_AIR DEST_AIR  SCHED_DEP  DEP_DELAY  \
0      1    1        4      WN     LAX      SLC       1625       58.0   
1      1    1        4      UA     DEN      IAD        823        7.0   
2      1    1        4      MQ     DFW      VPS       1305       36.0   
3      1    1        4      AA     DFW      DCA       1555        7.0   
4      1    1        4      WN     LAX      MCI       1720       48.0   

   AIR_TIME  DIST  SCHED_ARR  ARR_DELAY  DIVERTED  CANCELLED  
0      94.0   590       1905       65.0         0          0  
1     154.0  1452       1333      -13.0         0          0  
2      85.0   641       1453       35.0         0          0  
3     126.0  1192       1935       -7.0         0          0  
4     166.0  1363       2225       39.0         0          0

Now we want to have the mean of *ARR_DELAY* for each *AIRLINE*. We can accomplish this as the following: first we group by *AIRLINE* and then call the method **agg** which takes a dictionary in input specifying the column on what we want to apply the **mean** aggregate funcion.

In [8]:
flights.groupby('AIRLINE').agg({'ARR_DELAY': 'mean'}).head()

ARR_DELAY
AIRLINE           
AA        5.542661
AS       -0.833333
B6        8.692593
DL        0.339691
EV        7.034580

In [11]:
grouped = flights.groupby('AIRLINE')
type(grouped)

pandas.core.groupby.groupby.DataFrameGroupBy

Alternatively, it can be done as follows:

In [12]:
flights.groupby('AIRLINE')['ARR_DELAY'].agg('mean').head()

AIRLINE
AA    5.542661
AS   -0.833333
B6    8.692593
DL    0.339691
EV    7.034580
Name: ARR_DELAY, dtype: float64

You can pass directly the aggregate function as follows:

In [13]:
flights.groupby('AIRLINE')['ARR_DELAY'].agg(np.mean).head()

AIRLINE
AA    5.542661
AS   -0.833333
B6    8.692593
DL    0.339691
EV    7.034580
Name: ARR_DELAY, dtype: float64

The *agg* function can actually be skipped and obtain the same result:

In [14]:
flights.groupby('AIRLINE')['ARR_DELAY'].mean().head()

AIRLINE
AA    5.542661
AS   -0.833333
B6    8.692593
DL    0.339691
EV    7.034580
Name: ARR_DELAY, dtype: float64

## Grouping and aggregating with multiple columns and functions

In [4]:
flights = pd.read_csv('data/flights.csv')
flights.head()

MONTH  DAY  WEEKDAY AIRLINE ORG_AIR DEST_AIR  SCHED_DEP  DEP_DELAY  \
0      1    1        4      WN     LAX      SLC       1625       58.0   
1      1    1        4      UA     DEN      IAD        823        7.0   
2      1    1        4      MQ     DFW      VPS       1305       36.0   
3      1    1        4      AA     DFW      DCA       1555        7.0   
4      1    1        4      WN     LAX      MCI       1720       48.0   

   AIR_TIME  DIST  SCHED_ARR  ARR_DELAY  DIVERTED  CANCELLED  
0      94.0   590       1905       65.0         0          0  
1     154.0  1452       1333      -13.0         0          0  
2      85.0   641       1453       35.0         0          0  
3     126.0  1192       1935       -7.0         0          0  
4     166.0  1363       2225       39.0         0          0

**Question 1**: find the number of cancelled flights for every airline per day weekday

In [5]:
flights.groupby(['AIRLINE', 'WEEKDAY'])['CANCELLED'].sum().head(14)

AIRLINE  WEEKDAY
AA       1          41
         2           9
         3          16
         4          20
         5          18
         6          21
         7          29
AS       1           0
         2           0
         3           0
         4           0
         5           0
         6           0
         7           0
Name: CANCELLED, dtype: int64

**Question 2**: find the number and percentage of cancelled and diverted flights for every airline per weekday

In [27]:
flights.groupby(['AIRLINE', 'WEEKDAY'])['CANCELLED', 'DIVERTED'].agg(['sum', 'mean']).head(14)

CANCELLED           DIVERTED          
                      sum      mean      sum      mean
AIRLINE WEEKDAY                                       
AA      1              41  0.032106        6  0.004699
        2               9  0.007341        2  0.001631
        3              16  0.011949        2  0.001494
        4              20  0.015004        5  0.003751
        5              18  0.014151        1  0.000786
        6              21  0.018667        9  0.008000
        7              29  0.021837        1  0.000753
AS      1               0  0.000000        0  0.000000
        2               0  0.000000        0  0.000000
        3               0  0.000000        0  0.000000
        4               0  0.000000        0  0.000000
        5               0  0.000000        0  0.000000
        6               0  0.000000        0  0.000000
        7               0  0.000000        0  0.000000

**Question 3**: For each origin to destination flight, find the total number of flights, the number and percentage of cancelled flights and the average and variance of the airtime

In [10]:
group_cols = ['ORG_AIR', 'DEST_AIR']
agg_dict = {'CANCELLED' : ['sum', 'size', 'mean'], # NOTE: 'size' returns the total number of rows per group.
            'AIR_TIME' : ['mean', 'var']           # This is different from the 'count' aggregate function
           }                                       # because 'count' return the number of non missing values
flights.groupby(group_cols).agg(agg_dict).head(7)  # per group.

CANCELLED              AIR_TIME           
                       sum size mean        mean        var
ORG_AIR DEST_AIR                                           
ATL     ABE              0   31  0.0   96.387097  45.778495
        ABQ              0   16  0.0  170.500000  87.866667
        ABY              0   19  0.0   28.578947   6.590643
        ACY              0    6  0.0   91.333333  11.466667
        AEX              0   40  0.0   78.725000  47.332692
        AGS              0   83  0.0   28.819277   9.393770
        ALB              0   33  0.0  108.181818  41.903409

## Removing the MultiIndex after grouping

In [11]:
flights = pd.read_csv('data/flights.csv')
flights.head()

MONTH  DAY  WEEKDAY AIRLINE ORG_AIR DEST_AIR  SCHED_DEP  DEP_DELAY  \
0      1    1        4      WN     LAX      SLC       1625       58.0   
1      1    1        4      UA     DEN      IAD        823        7.0   
2      1    1        4      MQ     DFW      VPS       1305       36.0   
3      1    1        4      AA     DFW      DCA       1555        7.0   
4      1    1        4      WN     LAX      MCI       1720       48.0   

   AIR_TIME  DIST  SCHED_ARR  ARR_DELAY  DIVERTED  CANCELLED  
0      94.0   590       1905       65.0         0          0  
1     154.0  1452       1333      -13.0         0          0  
2      85.0   641       1453       35.0         0          0  
3     126.0  1192       1935       -7.0         0          0  
4     166.0  1363       2225       39.0         0          0

In [12]:
airline_info = flights.groupby(['AIRLINE', 'WEEKDAY'])\
                      .agg({'DIST':['sum', 'mean'], 
                                    'ARR_DELAY':['min', 'max']}).astype(int)
airline_info.head()

DIST       ARR_DELAY     
                     sum  mean       min  max
AIRLINE WEEKDAY                              
AA      1        1455386  1139       -60  551
        2        1358256  1107       -52  725
        3        1496665  1117       -45  473
        4        1452394  1089       -46  349
        5        1427749  1122       -41  732

Let's remove the MultiIndex. Use the method **get_level_values(level)** which accepts an integer identifying the index level

In [14]:
level0 = airline_info.columns.get_level_values(0)
level0

Index(['DIST', 'DIST', 'ARR_DELAY', 'ARR_DELAY'], dtype='object')

In [15]:
level1 = airline_info.columns.get_level_values(1)
level1

Index(['sum', 'mean', 'min', 'max'], dtype='object')

Now, we concatente both and display the result

In [17]:
airline_info.columns = level0 + '_' + level1
airline_info.head(7)

DIST_sum  DIST_mean  ARR_DELAY_min  ARR_DELAY_max
AIRLINE WEEKDAY                                                   
AA      1         1455386       1139            -60            551
        2         1358256       1107            -52            725
        3         1496665       1117            -45            473
        4         1452394       1089            -46            349
        5         1427749       1122            -41            732
        6         1265340       1124            -50            858
        7         1461906       1100            -49            626

Lastly, return the row labels to a single level using the **reset_index()**. Here we make the both index level as columns with *reset_index()*.

In [30]:
airline_info.reset_index().head(7)

AIRLINE  WEEKDAY  DIST_sum  DIST_mean  ARR_DELAY_min  ARR_DELAY_max
0      AA        1   1455386       1139            -60            551
1      AA        2   1358256       1107            -52            725
2      AA        3   1496665       1117            -45            473
3      AA        4   1452394       1089            -46            349
4      AA        5   1427749       1122            -41            732
5      AA        6   1265340       1124            -50            858
6      AA        7   1461906       1100            -49            626

Nice, we have successfully removed the MultiIndex.

## Examining a groupby object

In [41]:
college = pd.read_csv('data/college.csv')
grouped = college.groupby(['STABBR', 'RELAFFIL'])
type(grouped)

pandas.core.groupby.groupby.DataFrameGroupBy

Now we use the **dir** function to discover all its available functionality 

In [43]:
print([attr for attr in dir(grouped) if not attr.startswith('_')])

['CITY', 'CURROPER', 'DISTANCEONLY', 'GRAD_DEBT_MDN_SUPP', 'HBCU', 'INSTNM', 'MD_EARN_WNE_P10', 'MENONLY', 'PCTFLOAN', 'PCTPELL', 'PPTUG_EF', 'RELAFFIL', 'SATMTMID', 'SATVRMID', 'STABBR', 'UG25ABV', 'UGDS', 'UGDS_2MOR', 'UGDS_AIAN', 'UGDS_ASIAN', 'UGDS_BLACK', 'UGDS_HISP', 'UGDS_NHPI', 'UGDS_NRA', 'UGDS_UNKN', 'UGDS_WHITE', 'WOMENONLY', 'agg', 'aggregate', 'all', 'any', 'apply', 'backfill', 'bfill', 'boxplot', 'corr', 'corrwith', 'count', 'cov', 'cumcount', 'cummax', 'cummin', 'cumprod', 'cumsum', 'describe', 'diff', 'dtypes', 'expanding', 'ffill', 'fillna', 'filter', 'first', 'get_group', 'groups', 'head', 'hist', 'idxmax', 'idxmin', 'indices', 'last', 'mad', 'max', 'mean', 'median', 'min', 'ndim', 'ngroup', 'ngroups', 'nth', 'nunique', 'ohlc', 'pad', 'pct_change', 'pipe', 'plot', 'prod', 'quantile', 'rank', 'resample', 'rolling', 'sem', 'shift', 'size', 'skew', 'std', 'sum', 'tail', 'take', 'transform', 'tshift', 'var']


We can see that we have displayed all the public attribute and methods

Next, find the number of groups with the **ngroups** method

In [46]:
grouped.ngroups

112

Now, to find the uniquely identifying labels for each group, take the **keys()** dictionary wich contains ,for each group, a map to all corresponding index label of that group

In [51]:
groups = list(grouped.groups.keys())
groups[:6]

[('AK', 0), ('AK', 1), ('AL', 0), ('AL', 1), ('AR', 0), ('AR', 1)]

Now go ahead and retrieve a single group using the **get_group** method by passing a *tuple* of an exact group label

In [52]:
grouped.get_group(('FL', 1)).head()

INSTNM           CITY STABBR  HBCU  \
712         The Baptist College of Florida     Graceville     FL   0.0   
713                       Barry University          Miami     FL   0.0   
714  Gooding Institute of Nurse Anesthesia    Panama City     FL   0.0   
715             Bethune-Cookman University  Daytona Beach     FL   1.0   
724             Johnson University Florida      Kissimmee     FL   0.0   

     MENONLY  WOMENONLY  RELAFFIL  SATVRMID  SATMTMID  DISTANCEONLY  \
712      0.0        0.0         1     545.0     465.0           0.0   
713      0.0        0.0         1     470.0     462.0           0.0   
714      0.0        0.0         1       NaN       NaN           0.0   
715      0.0        0.0         1     405.0     395.0           0.0   
724      0.0        0.0         1     480.0     470.0           0.0   

            ...          UGDS_2MOR  UGDS_NRA  UGDS_UNKN  PPTUG_EF  CURROPER  \
712         ...             0.0308    0.0000     0.0507    0.2291         1   
713         ...             0.0164    0.0741     0.0841    0.1518         1   
714         ...                NaN       NaN        NaN       NaN         0   
715         ...             0.0198    0.0205     0.0190    0.0523         1   
724         ...             0.0045    0.0045     0.0136    0.1636         1   

     PCTPELL  PCTFLOAN  UG25ABV  MD_EARN_WNE_P10  GRAD_DEBT_MDN_SUPP  
712   0.5878    0.5602   0.3531            30800               20052  
713   0.5045    0.6733   0.4361            44100               28250  
714      NaN       NaN      NaN              NaN   PrivacySuppressed  
715   0.7758    0.8867   0.0647            29400               36250  
724   0.6689    0.7384   0.2185            26300               20199  

[5 rows x 27 columns]

You may want to take a peek of each individual group. This is possible because groupby objects are *iterable* 

Here we show a simple for loop for display each group

In [54]:
from IPython.display import display
i = 0
for name, group in grouped:
    print(name)
    display(group.head(2))
    i += 1
    if i == 5:
        break

('AK', 0)


INSTNM       CITY STABBR  HBCU  MENONLY  \
60  University of Alaska Anchorage  Anchorage     AK   0.0      0.0   
62  University of Alaska Fairbanks  Fairbanks     AK   0.0      0.0   

    WOMENONLY  RELAFFIL  SATVRMID  SATMTMID  DISTANCEONLY         ...          \
60        0.0         0       NaN       NaN           0.0         ...           
62        0.0         0       NaN       NaN           0.0         ...           

    UGDS_2MOR  UGDS_NRA  UGDS_UNKN  PPTUG_EF  CURROPER  PCTPELL  PCTFLOAN  \
60     0.0980    0.0181     0.0457    0.4539         1   0.2385    0.2647   
62     0.0401    0.0110     0.3060    0.3887         1   0.2263    0.2550   

    UG25ABV  MD_EARN_WNE_P10  GRAD_DEBT_MDN_SUPP  
60   0.4386            42500             19449.5  
62   0.4519            36200               19355  

[2 rows x 27 columns]

('AK', 1)


INSTNM       CITY STABBR  HBCU  MENONLY  WOMENONLY  \
61       Alaska Bible College     Palmer     AK   0.0      0.0        0.0   
64  Alaska Pacific University  Anchorage     AK   0.0      0.0        0.0   

    RELAFFIL  SATVRMID  SATMTMID  DISTANCEONLY         ...          UGDS_2MOR  \
61         1       NaN       NaN           0.0         ...             0.0370   
64         1     555.0     503.0           0.0         ...             0.0945   

    UGDS_NRA  UGDS_UNKN  PPTUG_EF  CURROPER  PCTPELL  PCTFLOAN  UG25ABV  \
61       0.0     0.0000    0.1481         1   0.3571    0.2857   0.4286   
64       0.0     0.0873    0.3745         1   0.3152    0.5297   0.4910   

    MD_EARN_WNE_P10  GRAD_DEBT_MDN_SUPP  
61              NaN   PrivacySuppressed  
64            47000               23250  

[2 rows x 27 columns]

('AL', 0)


INSTNM        CITY STABBR  HBCU  MENONLY  \
0             Alabama A & M University      Normal     AL   1.0      0.0   
1  University of Alabama at Birmingham  Birmingham     AL   0.0      0.0   

   WOMENONLY  RELAFFIL  SATVRMID  SATMTMID  DISTANCEONLY         ...          \
0        0.0         0     424.0     420.0           0.0         ...           
1        0.0         0     570.0     565.0           0.0         ...           

   UGDS_2MOR  UGDS_NRA  UGDS_UNKN  PPTUG_EF  CURROPER  PCTPELL  PCTFLOAN  \
0     0.0000    0.0059     0.0138    0.0656         1   0.7356    0.8284   
1     0.0368    0.0179     0.0100    0.2607         1   0.3460    0.5214   

   UG25ABV  MD_EARN_WNE_P10  GRAD_DEBT_MDN_SUPP  
0   0.1049            30300               33888  
1   0.2422            39700             21941.5  

[2 rows x 27 columns]

('AL', 1)


INSTNM        CITY STABBR  HBCU  MENONLY  WOMENONLY  \
2            Amridge University  Montgomery     AL   0.0      0.0        0.0   
10  Birmingham Southern College  Birmingham     AL   0.0      0.0        0.0   

    RELAFFIL  SATVRMID  SATMTMID  DISTANCEONLY         ...          UGDS_2MOR  \
2          1       NaN       NaN           1.0         ...             0.0000   
10         1     560.0     560.0           0.0         ...             0.0051   

    UGDS_NRA  UGDS_UNKN  PPTUG_EF  CURROPER  PCTPELL  PCTFLOAN  UG25ABV  \
2        0.0     0.2715    0.4536         1   0.6801    0.7795   0.8540   
10       0.0     0.0051    0.0017         1   0.1920    0.4809   0.0152   

    MD_EARN_WNE_P10  GRAD_DEBT_MDN_SUPP  
2             40100               23370  
10            44200               27000  

[2 rows x 27 columns]

('AR', 0)


INSTNM         CITY STABBR  HBCU  \
128        University of Arkansas at Little Rock  Little Rock     AR   0.0   
129  University of Arkansas for Medical Sciences  Little Rock     AR   0.0   

     MENONLY  WOMENONLY  RELAFFIL  SATVRMID  SATMTMID  DISTANCEONLY  \
128      0.0        0.0         0     470.0     510.0           0.0   
129      0.0        0.0         0       NaN       NaN           0.0   

            ...          UGDS_2MOR  UGDS_NRA  UGDS_UNKN  PPTUG_EF  CURROPER  \
128         ...             0.0755    0.0283     0.0003    0.4126         1   
129         ...             0.0281    0.0070     0.0169    0.2433         1   

     PCTPELL  PCTFLOAN  UG25ABV  MD_EARN_WNE_P10  GRAD_DEBT_MDN_SUPP  
128   0.3941    0.4775   0.4062            33900               21736  
129   0.3944    0.6144   0.5133            61400               12500  

[2 rows x 27 columns]

## Transforming through a weight-loss

In [55]:
weight_loss = pd.read_csv('data/weight_loss.csv')
weight_loss.query('Month == "Jan"')

Name Month    Week  Weight
0  Bob   Jan  Week 1     291
1  Amy   Jan  Week 1     197
2  Bob   Jan  Week 2     288
3  Amy   Jan  Week 2     189
4  Bob   Jan  Week 3     283
5  Amy   Jan  Week 3     189
6  Bob   Jan  Week 4     283
7  Amy   Jan  Week 4     190

To determine the winner for each month we only need to compare weight-loss from the firt week to the last week of each month.

With the following function we determine the percentage of weight-loss for each week comparing to the starter weight

In [56]:
def find_perc_loss(s):
    return (s - s.iloc[0]) / s.iloc[0]

In [60]:
bob_jan = weight_loss.query('Name=="Bob" and Month=="Jan"')
find_perc_loss(bob_jan['Weight'])

0    0.000000
2   -0.010309
4   -0.027491
6   -0.027491
Name: Weight, dtype: float64

We can see for example that after the first week of January, Bob lost *0.010309* of his initial body weight

We can apply this function to every single combination of person and week to get the weight-loss per week in relation to the **first** week of the month. To do this, we need to group our data by *Name* and *Month* and then use the **transform** method to apply our custom function.
With the following code, we are transforming the *Weight* column into the percentage of the weight-loss for the curret week: 

In [63]:
pcnt_loss = weight_loss.groupby(['Name', 'Month'])['Weight'].transform(find_perc_loss)
pcnt_loss.head(8)

0    0.000000
1    0.000000
2   -0.010309
3   -0.040609
4   -0.027491
5   -0.040609
6   -0.027491
7   -0.035533
Name: Weight, dtype: float64

It can be observed that the first month of data is outputted for each person. Pandas returns the new data as a Series. This Series is not so useful by itself, it make more sense by appending the result to the original datafram e as a new column:

In [64]:
weight_loss['Perc Weight Loss'] = pcnt_loss.round(3)
weight_loss.query('Name=="Bob" and Month in ["Jan", "Feb"]')

Name Month    Week  Weight  Perc Weight Loss
0   Bob   Jan  Week 1     291             0.000
2   Bob   Jan  Week 2     288            -0.010
4   Bob   Jan  Week 3     283            -0.027
6   Bob   Jan  Week 4     283            -0.027
8   Bob   Feb  Week 1     283             0.000
10  Bob   Feb  Week 2     275            -0.028
12  Bob   Feb  Week 3     268            -0.053
14  Bob   Feb  Week 4     268            -0.053

It can be observed in the output that the *Perc Weight Loss* resets after the new month.

Now with *Perc Weight Loss* column we can determine who's the winner. But let's see if we can do it automatically.
As the only week that matters is the *Week 4*, let's select it:

In [65]:
week4 = weight_loss.query('Week == "Week 4"')
week4

Name Month    Week  Weight  Perc Weight Loss
6   Bob   Jan  Week 4     283            -0.027
7   Amy   Jan  Week 4     190            -0.036
14  Bob   Feb  Week 4     268            -0.053
15  Amy   Feb  Week 4     173            -0.089
22  Bob   Mar  Week 4     261            -0.026
23  Amy   Mar  Week 4     170            -0.017
30  Bob   Apr  Week 4     250            -0.042
31  Amy   Apr  Week 4     161            -0.053

Now, let's reshape the data with the **pivot** method so that Bob's and Amy's Perc Weight Loss is side by side for each month

In [66]:
winner = week4.pivot(index='Month', columns='Name', values='Perc Weight Loss')
winner

Name     Amy    Bob
Month              
Apr   -0.053 -0.042
Feb   -0.089 -0.053
Jan   -0.036 -0.027
Mar   -0.017 -0.026

In the *pivot* method, **index** parameter is used for the column that you do **not** want to pivot. The columns passed to the **values** parameter gets titled of each unique combination of the columns in the **index** and **columns** parameter.

In the output is clear who have won each month but we can still go a couple of steps further. Numpy has a vectoriezed *if then else* function called **where** which can map a Series or an Array of booleans to other values 

In [76]:
winner['Winner'] = np.where(winner['Amy'] < winner['Bob'], 'Amy', 'Bob')
winner.style.highlight_min(axis=1)

The first argument (*winner['Amy'] < winner['Bob']*) is a condition that produces a Series of booleans. **True** values get mapped to **Amy** whereas **False** values get mapped to **Bob**.

Lastly, use the **value_counts()** method to return the final score

In [77]:
winner.Winner.value_counts()

Amy    3
Bob    1
Name: Winner, dtype: int64

We can now state that Amy won for three month and Bob won for onwe month. 